Adds get_follow_redirect validator method for browse-everything.

scherztc opened the pull request:
Fixes #1636, #1379 

Google Drive and Kaltura were presenting re-direct URIs.  These redirects were not parsing into the hyrax ImportURL job.  By running them through the get_follow_redirect validator, we get the proper uri for each file.

hortongn commented:
@scherztc Your commit includes `config/initializers/hydra_config.rb`.  Should that have been left out?  Also, coveralls is complaining that one line in the job isn't being tested: https://coveralls.io/jobs/29829481/source_files/1783567012#L17

scherztc commented:
@hortongn Coveralls is happy when I removed that config/initializers file.  I added that to the PR because it was a deprecation warning when I was running my test.

hortongn commented:
@scherztc But you still have an untested condition in your code.  Your tests handle cases where the response code is a 200 or a 302, but you don't have a test for when the response code is something else (like a 404).  So line 17 of your job never gets tested.

scherztc commented on app/jobs/import_url_job.rb:
```diff
@@ -0,0 +1,36 @@
+      when 301..303
+        get_follow_redirects(URI(response['location']), request_max - 1)
+      else
+        response.code
```
@hrotongn :  I don't know if I can test directly for a response.code in spec/jobs.  Everything I read says, that I can only use the have_http_status matcher in a controller, request, or feature spec.  I did add a success/failure test for the operation.  See below.

https://relishapp.com/rspec/rspec-rails/docs/matchers/have-http-status-matcher

hortongn replied:
@scherztc Line #17 is still a problem.  The tests don't cover it and if the application hits that line it will crash the app because calling `get_follow_redirects` will return a string like "500" or "404" instead of a uri.

If there's no chance the application will ever hit that else condition, we should just remove it.  However, if a code other than 200 or 301..303 is possible, the method needs to gracefully deal with it instead of just returning `response.code`

You can get a test to hit that else condition by stubbing a 500 response and then calling .perform now:

In the before block
`WebMock.stub_request(:get, URI("http://example.org#{file_hash}")).to_return(
  body: File.open(File.expand_path(file_path, __FILE__)).read, status: 500, headers: { 'Content-Type' => 'image/png', 'location' => URI("http://example.org") }
)`

in the test:
`described_class.perform_now(file_set, operation)`

It crashes with exception:
```
Failure/Error:
       retriever.retrieve(spec) do |chunk|
         f.write(chunk)
       end

     NameError:
       undefined local variable or method `uri' for #<BrowseEverything::Retriever:0x00000007db5950 @chunk_size=16384>
       Did you mean?  url
                      URI
```

scherztc replied:
@hortongn  I am looking into what Penn State did with Scholarsphere and trying to see if we can merge the two approaches.  https://github.com/psu-stewardship/scholarsphere/commit/22fe08c1957dda228bf4d6a5e239f6a48bee25f1#diff-27b3e499e0f5cc53eda5decc9879204d

scherztc commented on spec/jobs/import_url_job_spec.rb:
```diff
@@ -0,0 +1,108 @@
+    end
+
+    it 'does not return the url' do
+      expect(operation).not_to be_success
```
Since this is job spec, I added a success/failure test instead of a status_code test.

crowesn commented:
👍

hortongn commented:
@scherztc One simple approach is to just assume that if the response code is not a 200 or 301..303, just tell the user that something went wrong (rescue the error and flash an alert).  It's probably impossible to know and handle all of the situations given that so many things can go wrong.

So something like: 
```
  case response.code.to_i
  when 200
    uri
  when 301..303
    get_follow_redirects(URI(response['location']), request_max - 1)
  end
rescue NameError
  flash[:alert] = 'Something went wrong with adding the file.  Please check the file and try again"
```

One real-world situation where something like this can happen is when a user selects a file from their cloud provider, but before they save the work the file gets removed from their cloud drive.  So the URI that Scholar has now points to a nonexistent file and would probably generate a 404 or similar.

hortongn commented:
Actually this for the rescue instead:

```
retriever.retrieve(spec) do |chunk|
        f.write(chunk)
      end
      f.rewind
rescue NameError
  flash[:alert] = 'Something went wrong with adding the file.  Please check the file and try again"
end
```
